xapian-core/weight/tfidfweight.cc

```diff
@@ -76,12 +78,18 @@ TfIdfWeight::TfIdfWeight(const std::string &normals, double slope, double delta)
     need_stat(WDF);
     need_stat(WDF_MAX);
     need_stat(WQF);
-    need_stat(DOC_LENGTH);
-    need_stat(DOC_LENGTH_MIN);
-    need_stat(DOC_LENGTH_MAX);
-    need_stat(UNIQUE_TERMS);
     if (normalizations[0] == 'P' || normalizations[1] == 'P') {
 	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(DOC_LENGTH_MIN);
+    }
+    else if (normalizations[0] == 'L') {
+	need_stat(DOC_LENGTH);
+	need_stat(DOC_LENGTH_MIN);
+	need_stat(DOC_LENGTH_MAX);
+	need_stat(UNIQUE_TERMS);
+	if (normalizations[1] == 'P')
+	    need_stat(AVERAGE_LENGTH);
```

Comment:
But as I noted above, `need_stat()` just does a bitwise or, it doesn't go and fetch or calculate the statistic. So those "extra duplicate calls" only means the bit for `AVERAGE_LENGTH` is set in `SOME_CONSTANT1` and `SOME_CONSTANT2`.

Once the compiler has inlined and optimised the code, from an efficiency point of view what matters is not the number of calls to `need_stat()` in the source code, but how many separate runs of such calls.